9.2/run-postgresql.sh

```diff
@@ -98,6 +109,14 @@ function initialize_database() {
 # additional custom settings and is included from $PGDATA/postgresql.conf.
 envsubst < ${POSTGRESQL_CONFIG_FILE}.template > ${POSTGRESQL_CONFIG_FILE}
 
+# Generate passwd file based on current uid
+export USER_ID=$(id -u)
+envsubst < ${HOME}/passwd.template > ${HOME}/passwd
+cat ${HOME}/passwd
+export LD_PRELOAD=libnss_wrapper.so
+export NSS_WRAPPER_PASSWD=/var/lib/pgsql/passwd
+export NSS_WRAPPER_GROUP=/etc/group
```

Comment:
In the 9.2 source base, this is the code in initdb.c:
```c
static char *
get_id(void)
{
#ifndef WIN32

	struct passwd *pw;

	if (geteuid() == 0)			/* 0 is root's uid */
	{
		fprintf(stderr,
				_("%s: cannot be run as root\n"
				  "Please log in (using, e.g., \"su\") as the "
				  "(unprivileged) user that will\n"
				  "own the server process.\n"),
				progname);
		exit(1);
	}

	pw = getpwuid(geteuid());
	if (!pw)
	{
		fprintf(stderr,
			  _("%s: could not obtain information about current user: %s\n"),
				progname, strerror(errno));
		exit(1);
	}
```
In later versions, it's been replaced by another function (get_username_or_exit) but it's the same deal:
http://doxygen.postgresql.org/username_8c.html#a0e84f99bf0c1826e3f629131a7719aac